bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/file_processor_nwd.cpp

```diff
@@ -0,0 +1,821 @@
+
+#include <Attribute/NwAttribute.h>
+#include <Attribute/NwPropertyAttribute.h>
+#include <Attribute/NwGraphicMaterialAttribute.h>
+#include <Attribute/NwGuidAttribute.h>
+#include <Attribute/NwMaterialAttribute.h>
+#include <Attribute/NwNameAttribute.h>
+#include <Attribute/NwBinaryAttribute.h>
+#include <Attribute/NwPresenterMaterialAttribute.h>
+#include <Attribute/NwPresenterTextureSpaceAttribute.h>
+#include <Attribute/NwPublishAttribute.h>
+#include <Attribute/NwTextAttribute.h>
+#include <Attribute/NwTransAttribute.h>
+#include <Attribute/NwTransformAttribute.h>
+#include <Attribute/NwTransRotationAttribute.h>
+#include <Attribute/NwUInt64Attribute.h>
+#include <Attribute/NwURLAttribute.h>
+
+//3d repo bouncer
+#include "file_processor_nwd.h"
+#include "helper_functions.h"
+
+//boost
+#include <boost/tokenizer.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/filesystem.hpp>
+
+using namespace repo::manipulator::modelconvertor::odaHelper;
+
+static OdString sNwDbModuleName = L"TNW_Db";
+static std::string sElementIdKey = "Element ID::Value";
+static std::vector<std::string> ignoredCategories = { "Material", "Geometry", "Autodesk Material", "Revit Material" };
+
+class OdExNwSystemServices : public ExSystemServices
+{
+public:
+	OdExNwSystemServices() {}
+};
+
+class RepoNwServices : public OdExNwSystemServices, public OdNwHostAppServices
+{
+protected:
+	ODRX_USING_HEAP_OPERATORS(OdExNwSystemServices);
+};
+
+repo::manipulator::modelconvertor::odaHelper::FileProcessorNwd::~FileProcessorNwd()
+{
+}
+
+struct RepoNwTraversalContext {
+	OdNwModelItemPtr layer;
+	OdNwPartitionPtr partition;
+	std::string parent;
+	GeometryCollector* collector;
+};
+
+repo::lib::RepoVector3D64 convertPoint(OdGePoint3d pnt, OdGeMatrix3d& transform)
+{
+	pnt.transformBy(transform);
+	return repo::lib::RepoVector3D64(pnt.x, pnt.y, pnt.z);
+};
+
+repo::lib::RepoVector3D64 convertPoint(OdGePoint3d pnt) 
+{ 
+	return repo::lib::RepoVector3D64(pnt.x, pnt.y, pnt.z); 
+};
+
+repo::lib::RepoVector2D convertPoint(OdGePoint2d pnt)
+{
+	return repo::lib::RepoVector2D(pnt.x, pnt.y);
+};
+
+void convertColor(OdNwColor color, std::vector<float>& dest)
+{
+	dest.clear();
+	dest.push_back(color.R());
+	dest.push_back(color.G());
+	dest.push_back(color.B());
+	dest.push_back(color.A());
+}
+
+void convertColor(OdString color, std::vector<float>& dest)
+{
+	dest.clear();
+	try 
+	{
+		typedef boost::tokenizer<boost::char_separator<OdChar>> tokenizer;
+		boost::char_separator<OdChar> sep(OD_T(","));
+		auto stringified = convertToStdString(color);
+		tokenizer tokens(stringified, sep);
+		for (tokenizer::iterator iter = tokens.begin(); iter != tokens.end(); iter++) {
+			auto token = *iter;
+			boost::trim(token);
+			auto component = boost::lexical_cast<float>(token);
+			dest.push_back(component / 255.0);
+		}
+	}
+	catch(...)
+	{
+		repoError << "Exception parsing string " << convertToStdString(color) << " to a color.";
```

Comment:
do we expect this to happen at all?

Reply:
It is not expected, but we have seen something similar happen in the RVT importer. I added it mostly because if the exception isn't caught here it goes all the way up to the file processor and kills the whole import.